core/src/main/scala/kafka/server/AlterIsrManager.scala

```diff
@@ -122,43 +121,47 @@ class DefaultAlterIsrManager(
 
   override def start(): Unit = {
     controllerChannelManager.start()
-    scheduler.schedule("send-alter-isr", propagateIsrChanges, 50, 50, TimeUnit.MILLISECONDS)
   }
 
   override def shutdown(): Unit = {
     controllerChannelManager.shutdown()
   }
 
   override def submit(alterIsrItem: AlterIsrItem): Boolean = {
-    unsentIsrUpdates.putIfAbsent(alterIsrItem.topicPartition, alterIsrItem) == null
+    if (unsentIsrUpdates.putIfAbsent(alterIsrItem.topicPartition, alterIsrItem) == null) {
+      if (inflightRequest.compareAndSet(false, true)) {
```

Comment:
This seems to introduce a race condition. Say we have an inflight request. Is the following sequence possible?

1. the response returns and the io thread calls `propagateIsrChanges` and sees an empty `unsentIsrUpdates`
2. request thread calls submit and inserts a new item in `unsentIsrUpdates`
3. request thread fails `compareAndSet` on `inflightRequest`
4. io thread clears `inflightRequest`

It seems like we might need a lock.

Reply:
You're right, there's a race between the isEmpty check and clearing the inflight flag. Good catch